Implement RS485 support (add TX_ENABLE signal)

deece opened the pull request:
Once accepted, this will implement the feature request in #55.

This is currently untested (I'm waiting for boards to be fabbed, could take ~3 weeks), but some code review in the meantime would be appreciated.

This patch adds a TX_ENABLE pin to the GUI (default is unassigned). When enabled, the signal is asserted in uart.c:uart_tx_one_char. The signal is deasserted through the TX_FIFO_EMPTY interrupt, followed by a timer to ensure the last character read from the buffer has finished transmission.

The os timer has been switched from millisecond to microsecond mode to allow this.

deece committed:
Add support for the TX_ENABLE signal for RS-485 trancievers

deece committed:
Use microsecond timer for heap use as the timer is in microsecond mode

deece committed:
Update all timer instances to microsecond timers

tve commented on esp-link/cgiflash.c:
```diff
@@ -13,6 +13,7 @@ Some flash handling cgi routines. Used for reading the existing flash and updati
  * ----------------------------------------------------------------------------
  */
 
+#define USE_US_TIMER
```
Shouldn't this go into a .h file so it gets included everywhere?
Note to self: this limits the timer to 268s max

tve commented on serial/uart.c:
```diff
@@ -81,12 +130,19 @@ uart_config(uint8 uart_no)
+
+	uint32_t tx_empty_bits = 0;
+	if (uart0_tx_enable_pin >= 0) {
+	  tx_empty_bits = (0 & UART_TXFIFO_EMPTY_THRHD) << UART_TXFIFO_EMPTY_THRHD_S | UART_TXFIFO_EMPTY_INT_ENA;
```
I'm not sure I understand... Isn't the value of tx_empty_bits always 0 due to `(0 & UART_TXFIFO_EMPTY_THRHD)`?

deece replied:
It's UART_TXFIFO_EMPTY_INT_ENA, the first part evaluates to 0, but I left it there to show that we are setting thethreshold to 0.

tve commented on serial/uart.c:
```diff
@@ -110,6 +169,16 @@ uart_tx_one_char(uint8 uart, uint8 c)
+	}
+
+    tx_enable(true);
+  }
```
Shouldn't this code be conditional on the tx_enable_pin feature being enabled? It seems to set the timer unconditionally.

Also, I believe this needs to be done with interrupts disabled, else there's a race condition if the interrupt hits between `uart_tx_enable_timer_inited` and  `os_timer_setfn`.

deece replied:
I think that situation should be safe, as the timer interrupt should only fire once the last character has finished transmitting. This would cause TX_ENABLE to go low, and it will then be set high very soon after by the tx_enable(true) call, before the UART is activated again.

tve commented on serial/uart.c:
```diff
@@ -206,6 +276,12 @@ uart0_rx_intr_handler(void *para)
+	if (0 != uart0_baud_rate) {
+	  int tx_char_time = 8 * 1000000 / uart0_baud_rate; // assumes 8 bits per character
+	  os_timer_arm_us(&uart_tx_enable_timer, tx_char_time, false);
+	}
```
shouldn't this set `uart_tx_enable_timer_inited`?

whenever calling os_timer_arm_us you should disable it first, just in case. See SDK documentation:
- if called system_timer_reinit, the maximum value allowed to input is
  0x41893
- if didn’t call system_timer_reinit, the maximum value allowed to input
  is 0xFFFFFFF

deece committed:
Move USE_US_TIMER to esp8266.h and remove redundant inclusions of osapi.h

deece committed:
Add a clarifying comment about the setup of TX empty

deece committed:
Ensure the TX timer is disarmed after it has fired, allowing it to be retriggered.
Only manipulate the timer if TX_ENABLE has been requested.
Assert uart_tx_enable_timer_inited once the timer has been inited.

deece committed:
Add missing initialisation for uart0_tx_enable_pin

krzy27 commented on esp-link/cgipins.c:
```diff
@@ -32,9 +32,10 @@ int ICACHE_FLASH_ATTR cgiPinsGet(HttpdConnData *connData) {
 
   len = os_sprintf(buff,
-      "{ \"reset\":%d, \"isp\":%d, \"conn\":%d, \"ser\":%d, \"swap\":%d, \"rxpup\":%d }",
+      "{ \"reset\":%d, \"isp\":%d, \"conn\":%d, \"ser\":%d, \"swap\":%d, \"rxpup\":%d, \"uart0_tx_enable\":%d }",
```
`"{ \"reset\":%d, \"isp\":%d, \"conn\":%d, \"ser\":%d, \"swap\":%d, \"rxpup\":%d, \"txe\":%d }",`
I prefer short name `txe` for html request.

krzy27 commented on esp-link/cgipins.c:
```diff
@@ -48,14 +49,15 @@ int ICACHE_FLASH_ATTR cgiPinsSet(HttpdConnData *connData) {
   ok |= getInt8Arg(connData, "ser", &ser);
   ok |= getBoolArg(connData, "swap", &swap);
   ok |= getBoolArg(connData, "rxpup", &rxpup);
+  ok |= getInt8Arg(connData, "uart0_tx_enable", &uart0_tx_enable);
```
`ok |= getInt8Arg(connData, "txe", &uart0_tx_enable);`

krzy27 commented on esp-link/cgipins.c:
```diff
@@ -90,8 +96,8 @@ int ICACHE_FLASH_ATTR cgiPinsSet(HttpdConnData *connData) {
     flashConfig.ser_led_pin = ser;
     flashConfig.swap_uart = swap;
     flashConfig.rx_pullup = rxpup;
```
add line:
`flashConfig.uart0_tx_enable_pin = uart0_tx_enable;`

krzy27 commented on html/home.html:
```diff
@@ -77,6 +77,11 @@ <h2 id="version"></h2>
                     AVR:not used, esp8266:gpio2, ARM:ISP</div>
                 </div>
                 <div class="pure-control-group">
+                  <label for="pin-uart0-tx-enable">TX Enable</label>
```
`<label for="pin-txe">TX Enable</label>`

krzy27 commented on html/home.html:
```diff
@@ -77,6 +77,11 @@ <h2 id="version"></h2>
                 </div>
                 <div class="pure-control-group">
+                  <label for="pin-uart0-tx-enable">TX Enable</label>
+                  <select id="pin-uart0-tx-enable"></select>
```
`<select id="pin-txe"></select>`

krzy27 commented on html/ui.js:
```diff
@@ -418,6 +418,7 @@ function createPresets(sel) {
     setPP("ser",   pp[3]);
     setPP("swap",  pp[4]);
     $("#pin-rxpup").checked = !!pp[5];
+    setPP("uart0-tx-enable",  pp[6]);
```
`setPP("txe",  pp[6]);`

krzy27 commented on html/ui.js:
```diff
@@ -453,6 +454,7 @@ function displayPins(resp) {
   createSelectForPin("ser", resp["ser"]);
   $("#pin-swap").value = resp["swap"];
   $("#pin-rxpup").checked = !!resp["rxpup"];
+  createSelectForPin("uart0_tx_enable", resp["uart0_tx_enable"]);
```
`createSelectForPin("txe", resp["txe"]);`

krzy27 commented on html/ui.js:
```diff
@@ -469,7 +471,7 @@ function setPins(ev) {
   var url = "/pins";
   var sep = "?";
-  ["reset", "isp", "conn", "ser", "swap"].forEach(function(p) {
+  ["reset", "isp", "conn", "ser", "swap", "uart0_tx_enable"].forEach(function(p) {
```
`["reset", "isp", "conn", "ser", "swap", "txe"].forEach(function(p) {`

deece committed:
Standardize on uart0-tx-enable variable name on the HTML side (thanks @krzy27)

deece committed:
Fix typo

krzy27 commented on serial/serbridge.c:
```diff
@@ -463,9 +464,17 @@ serbridgeInitPins()
   if (mcu_isp_pin >= 0)   GPIO_OUTPUT_SET(mcu_isp_pin, 1);
   if (mcu_reset_pin >= 0) GPIO_OUTPUT_SET(mcu_reset_pin, 1);
+  // set TX_ENABLE to 0 so we start up listening
+  if (mcu_reset_pin >= 0) GPIO_OUTPUT_SET(uart0_tx_enable_pin, 0);
```
`if (uart0_tx_enable_pin >= 0) GPIO_OUTPUT_SET(uart0_tx_enable_pin, 0);`
Start listening even reset pin is disable.

krzy27 commented on serial/uart.c:
```diff
@@ -37,6 +38,56 @@ static UartRecv_cb uart_recv_cb[4];
+ * Returns      : NONE
+*******************************************************************************/
+static void ICACHE_FLASH_ATTR
+tx_completed_interrupt(void *unused)
```
Something is wrong witch this interrupt. If esp start transmit newer come back to listening.

deece committed:
Fix tx_enable init logic
save a couple of branches

deece committed:
fix variable name

krzy27 commented on serial/uart.c:
```diff
@@ -81,12 +132,20 @@ uart_config(uint8 uart_no)
+                   UART_RX_TOUT_EN) |
+                   tx_empty_bits
+         ;
     SET_PERI_REG_MASK(UART_INT_ENA(uart_no), UART_RXFIFO_FULL_INT_ENA | UART_RXFIFO_TOUT_INT_ENA);
```
If I change `SET_PERI_REG_MASK(UART_INT_ENA(uart_no), UART_RXFIFO_FULL_INT_ENA | UART_RXFIFO_TOUT_INT_ENA | UART_TXFIFO_EMPTY_INT_ENA);` then program continuously going to line 282 and make error "unable run/load timer" or some like this. After few second watch dog make reset.

deece committed:
Apply @krzy27's changes to set up the timers correctly.

deece committed:
Merge remote-tracking branch 'origin/master' into rs485

deece committed:
Merge branch 'master' into rs485

deece committed:
Fix bad merge

Add missing syntax

deece committed:
New flash entries should be added at the end

deece committed:
New flash entries should be added at the end
